# Git's Objects

- git's object storage is
    - key value storage
    - key is SHA-1 hash of object, with some additional information like size

- 4 types of objects are there, they are reffered by
    - branches
    - HEAD

## 4 Types of git objects

- tree (directories)
- blobs (files)
- commits
- tags


In [4]:
pwd

/home/ncoder/02_wos/jupyter-lab/notebooks/03_git/02_git-mastering-version-control-sfrln/2-git-version-control-cookbook


In [5]:
cd ../2-2-git-version-control-cookbook

In [6]:
pwd

/home/ncoder/02_wos/jupyter-lab/notebooks/03_git/02_git-mastering-version-control-sfrln/2-2-git-version-control-cookbook


In [7]:
git clone https://github.com/dvaske/data-model.git

Cloning into 'data-model'...
remote: Enumerating objects: 45, done.
remote: Counting objects: 100% (45/45), done.
remote: Compressing objects: 100% (26/26), done.
remote: Total 45 (delta 11), reused 45 (delta 11), pack-reused 0
Unpacking objects: 100% (45/45), done.


In [8]:
ls

 c1.navigating-git.ipynb  'c2.gits objects.ipynb'   data-model   images


In [11]:
cd data-model;
pwd

bash: cd: data-model: No such file or directory
/home/ncoder/02_wos/jupyter-lab/notebooks/03_git/02_git-mastering-version-control-sfrln/2-2-git-version-control-cookbook/data-model


## Looking at COMMIT object

### HEAD points to
- current snapshot/commit

#### command used to look at a COMMIT

- `git cat-file`

In [12]:
git cat-file -p HEAD

tree 34fa038544bcd9aed660c08320214bafff94150b
parent a90d1906337a6d75f1dc32da647931f932500d83
author Aske Olsson <aske.olsson@switch-gears.dk> 1386933960 +0100
committer Aske Olsson <aske.olsson@switch-gears.dk> 1386941455 +0100

This is the subject line of the commit message

It should be followed by a blank line then the body, which is this text. Here
you can have multiple paragraphs etc. and explain your commit. It's like an
email with subject and body, so get people's attention in the subject


#### git cat-file output

- `cat-file -p` pretty prints given object
    - in this case, HEAD which points to master branch, which inturn points to latest commit on master branch
- what does commit object consists of?
    - root tree (tree 34fa038544bcd9aed660c08320214bafff94150b)
    - parent commit object's id (parent a90d1906337a6d75f1dc32da647931f932500d83)
    - author and timestamp information (author Aske Olsson <aske.olsson@switch-gears.dk> 1386933960 +0100)
    - committer and timestamp information (committer Aske Olsson <aske.olsson@switch-gears.dk> 1386941455 +0100)
    - commit message
    

### differnce between git commit author and committer

The author is the person who originally wrote the code. The committer, on the other hand, is assumed to be the person who committed the code on behalf of the original author. This is important in Git because Git allows you to rewrite history, or apply patches on behalf of another person.

ref: https://stackoverflow.com/questions/18750808/difference-between-author-and-committer-in-git

## Looking into TREE object

**How**
- using `git cat-file -p` command with the tree ID as the argument

In [15]:
git cat-file -p 34fa038544bcd9aed660c08320214bafff94150b

100644 blob f21dc2804e888fee6014d7e5b1ceee533b222c15	README.md
040000 tree abc267d04fb803760b75be7e665d3d69eeed32f8	a_sub_directory
100644 blob b50f80ac4d0a36780f9c0636f43472962154a11a	another-file.txt
100644 blob 92f046f17079aa82c924a9acf28d623fcb6ca727	cat-me.txt
100644 blob bb2fe940924c65b4a1cefcbdbe88c74d39eb23cd	hello_world.c


### git cat-file 

In [16]:
git cat-file -h

usage: git cat-file (-t [--allow-unknown-type] | -s [--allow-unknown-type] | -e | -p | <type> | --textconv | --filters) [--path=<path>] <object>
   or: git cat-file (--batch | --batch-check) [--follow-symlinks] [--textconv | --filters]

<type> can be one of: blob, tree, commit, tag
    -t                    show object type
    -s                    show object size
    -e                    exit with zero when there's no error
    -p                    pretty-print object's content
    --textconv            for blob objects, run textconv on object's content
    --filters             for blob objects, run filters on object's content
    --path <blob>         use a specific path for --textconv/--filters
    --allow-unknown-type  allow -s and -t to work with broken/corrupt objects
    --buffer              buffer --batch output
    --batch[=<format>]    show info and content of objects fed from the standard input
    --batch-check[=<format>]
                          show info about obje

: 129

In [17]:
git cat-file -t 34fa038544bcd9aed660c08320214bafff94150b

tree


In [18]:
git cat-file -s 34fa038544bcd9aed660c08320214bafff94150b

202


### Looking into the tree object of a Rev

the tree object of a HEAD can be printed with
- `git cat-file -p HEAD^{tree} 

In [2]:
git cat-file -p HEAD^{tree}

100644 blob 8b137891791fe96927ad78e64b0aad7bded08bdc	.gitignore
100644 blob c84daf56fa2d5452ba1cd69fa67e9784f380f09d	.gitmodules
100644 blob 177aafe62f3951e9ac34905658b27a1f2e80dd48	Pipfile
100644 blob dcb4a125e6eb090179c0ea6b85f7c9455fe9e1ef	Pipfile.lock
040000 tree c2a269137439f20df79db350b07addd1f2438233	notebooks
100644 blob ee275c09328b70dbd12339cc532c2a8ebd492741	reference.md
100755 blob cfc06a524e0b4d1d56a7651f5f3dee388209b347	run.sh


#### generic form

`git cat-file -p <rev>^{type}`

means

- git will return first object of type `type` searching recursiveely from `rev`

#### cat-file tree object shows

- file type/permissions
- type (tree/blob)
- ID (SHA-1)
- pathname

## Looking into BLOB Object

display blob
- cat-file blob id

In [4]:
cd ../2-2-git-version-control-cookbook/data-model

In [5]:
git cat-file -p HEAD

tree 34fa038544bcd9aed660c08320214bafff94150b
parent a90d1906337a6d75f1dc32da647931f932500d83
author Aske Olsson <aske.olsson@switch-gears.dk> 1386933960 +0100
committer Aske Olsson <aske.olsson@switch-gears.dk> 1386941455 +0100

This is the subject line of the commit message

It should be followed by a blank line then the body, which is this text. Here
you can have multiple paragraphs etc. and explain your commit. It's like an
email with subject and body, so get people's attention in the subject


In [6]:
git cat-file -p 34fa038544bcd9aed660c08320214bafff94150b

100644 blob f21dc2804e888fee6014d7e5b1ceee533b222c15	README.md
040000 tree abc267d04fb803760b75be7e665d3d69eeed32f8	a_sub_directory
100644 blob b50f80ac4d0a36780f9c0636f43472962154a11a	another-file.txt
100644 blob 92f046f17079aa82c924a9acf28d623fcb6ca727	cat-me.txt
100644 blob bb2fe940924c65b4a1cefcbdbe88c74d39eb23cd	hello_world.c


In [7]:
git cat-file -p 92f046f17079aa82c924a9acf28d623fcb6ca727

This is the content of the file: "cat-me.txt"

Not really that exciting, huh?


### How objects are tied together
- blobs to trees
- trees to other trees
- root tree to the commit object
- all by the SHA-identifier of the object

## Looking into a branch

- cat-file will print the commit object a branch points to

In [8]:
git cat-file -p master

tree 34fa038544bcd9aed660c08320214bafff94150b
parent a90d1906337a6d75f1dc32da647931f932500d83
author Aske Olsson <aske.olsson@switch-gears.dk> 1386933960 +0100
committer Aske Olsson <aske.olsson@switch-gears.dk> 1386941455 +0100

This is the subject line of the commit message

It should be followed by a blank line then the body, which is this text. Here
you can have multiple paragraphs etc. and explain your commit. It's like an
email with subject and body, so get people's attention in the subject


### finding details about a branch

- we can find it inside .git folder where the whole git repo is stored
- text file .git/refs/<branch name> show commit the branch points to

In [9]:
cat .git/refs/heads/master

34acc370b4d6ae53f051255680feaefaf7f7850d


#### how to verify .git/refs/heads/master is pointing to the right commit
- use git log command

In [10]:
git log -1

commit 34acc370b4d6ae53f051255680feaefaf7f7850d (HEAD -> master, tag: v1.0, origin/master, origin/HEAD)
Author: Aske Olsson <aske.olsson@switch-gears.dk>
Date:   Fri Dec 13 12:26:00 2013 +0100

    This is the subject line of the commit message
    
    It should be followed by a blank line then the body, which is this text. Here
    you can have multiple paragraphs etc. and explain your commit. It's like an
    email with subject and body, so get people's attention in the subject


#### find branch pointed by HEAD

- cat .git/HEAD

In [11]:
cat .git/HEAD

ref: refs/heads/master


## Looking into the tag Object

### types of tag objects
- lightweight tag (just label)
- annotated tag
- signed tag

#### list tags
- use `git tag` command

In [12]:
git tag

v0.1
v1.0


#### what is inside a tag object
- can display using `git cat-file -p`

In [13]:
git cat-file -p v1.0

object 34acc370b4d6ae53f051255680feaefaf7f7850d
type commit
tag v1.0
tagger Aske Olsson <aske.olsson@switch-gears.dk> 1386941492 +0100

We got the hello world C program merged, let's call that a release 1.0


#### a tag consists of
- an object (in the above case it is the latest commit on master
- objects type (commits, blobs, trees can be tagged)
- tag name
- tagger
- tag message

## How to verify ouput of git cat-file
- by rehashing with `git hash-object` command


In [14]:
git cat-file -p HEAD

tree 34fa038544bcd9aed660c08320214bafff94150b
parent a90d1906337a6d75f1dc32da647931f932500d83
author Aske Olsson <aske.olsson@switch-gears.dk> 1386933960 +0100
committer Aske Olsson <aske.olsson@switch-gears.dk> 1386941455 +0100

This is the subject line of the commit message

It should be followed by a blank line then the body, which is this text. Here
you can have multiple paragraphs etc. and explain your commit. It's like an
email with subject and body, so get people's attention in the subject


In [15]:
git cat-file -p HEAD | git hash-object -t commit --stdin

34acc370b4d6ae53f051255680feaefaf7f7850d


In [16]:
git log -1

commit 34acc370b4d6ae53f051255680feaefaf7f7850d (HEAD -> master, tag: v1.0, origin/master, origin/HEAD)
Author: Aske Olsson <aske.olsson@switch-gears.dk>
Date:   Fri Dec 13 12:26:00 2013 +0100

    This is the subject line of the commit message
    
    It should be followed by a blank line then the body, which is this text. Here
    you can have multiple paragraphs etc. and explain your commit. It's like an
    email with subject and body, so get people's attention in the subject


## Other commands to see git objects in git database

- git show
- git ls-tree (show contentsof git trees and subtrees)

In [17]:
git show -h

usage: git log [<options>] [<revision-range>] [[--] <path>...]
   or: git show [<options>] <object>...

    -q, --quiet           suppress diff output
    --source              show source
    --use-mailmap         Use mail map file
    --decorate-refs <pattern>
                          only decorate refs that match <pattern>
    --decorate-refs-exclude <pattern>
                          do not decorate refs that match <pattern>
    --decorate[=...]      decorate options
    -L <n,m:file>         Process line range n,m in file, counting from 1



: 129

In [18]:
git ls-tree -h

usage: git ls-tree [<options>] <tree-ish> [<path>...]

    -d                    only show trees
    -r                    recurse into subtrees
    -t                    show trees when recursing
    -z                    terminate entries with NUL byte
    -l, --long            include object size
    --name-only           list only filenames
    --name-status         list only filenames
    --full-name           use full path names
    --full-tree           list entire tree; not just current directory (implies --full-name)
    --abbrev[=<n>]        use <n> digits to display SHA-1s



: 129

#### look into a HEAD using using different command than cat-file

In [21]:
git ls-tree HEAD

100644 blob f21dc2804e888fee6014d7e5b1ceee533b222c15	README.md
040000 tree abc267d04fb803760b75be7e665d3d69eeed32f8	a_sub_directory
100644 blob b50f80ac4d0a36780f9c0636f43472962154a11a	another-file.txt
100644 blob 92f046f17079aa82c924a9acf28d623fcb6ca727	cat-me.txt
100644 blob bb2fe940924c65b4a1cefcbdbe88c74d39eb23cd	hello_world.c


#### ls-tree show only trees

In [20]:
git ls-tree -d HEAD

040000 tree abc267d04fb803760b75be7e665d3d69eeed32f8	a_sub_directory


#### ls-tree show only tree, recursively

In [22]:
git ls-tree -d -r HEAD

040000 tree abc267d04fb803760b75be7e665d3d69eeed32f8	a_sub_directory
